src/coap_gnutls.c

```diff
@@ -1481,6 +1504,32 @@ coap_dtls_hello(coap_session_t *c_session,
   int ret;
 
   if (!g_env) {
+    /*
+     * Need to check that this actually is a Client Hello before wasting
+     * time allocating and then freeing off g_env.
+     */
+    const dtls_record_handshake_t *header =
```

Comment:
This construct will probably run into byte-order issues. As only `data[offsetof(dtls_record_handshake_t, content_type)]` and `data[offsetof(dtls_record_handshake_t, handshake)]` are used, there is no need for this type cast.

More important, I wonder if we really need to deal with the DTLS packet structure here? In theory, the DTLS library should at first decline any ClientHello with a HelloVerify (which GnuTLS here for some reason does not).

Reply:
Hmm,  whether using `header->content_type` or
`data[offsetof(dtls_record_handshake_t, content_type)]`  if there are going to be byte ordering issues (which should not be the case), then both methods will be susceptible to this. I have however made this change.

Originally, I did not have the pre-sanity checking that it was a ClientHello handshake which meant that g_env was getting built, GnuTLS library then said this was not a ClientHello handshake and g_env was then torn down again (my original fix attempt which worked as expected).

However, there is a significant performance requirement to go through this process, and it would not be that difficult to DOS the server by sending packets that are not ClientHello.  After some thinking, I decided to go with the packet outline sanity checking before setting up g_env etc.  I agree though that is it not an elegant part of the solution.

Reply:
> Hmm, whether using `header->content_type` or
> `data[offsetof(dtls_record_handshake_t, content_type)]` if there are going to be byte ordering issues (which should not be the case), then both methods will be susceptible to this. I have however made this change.

Sorry, I did not mean to literally use `offsetof` — I was too lazy to count (should be `0`, and `13`, I suppose).

Reply:
Code updated and pushed to use 0 and 13.